### ISSUE TITLE
Decomissioning plan for sdl1

### ISSUE THREAD

**Issue by lynxlynxlynx** [gemrb/gemrb#821]
Initially when discussing #441 last year I was sceptical of just dumping sdl1 for several reasons:
- it was the more mature backend on master
- it seemed work upstream continued still
- it would provide a nice base to test sdl2 experimentation against
- it'd be nice to warn people before completely dropping it

Focusing more on subviews with sdl1 again, I see we have several bugs, some whack-a-moly, so perhaps it's better to dump it sooner than later. Nobody is doing 1.2.16 either, commits are just piling up.

Considering everything, I propose we do the anniversary release from master, which is just shy of getting the finished pathfinder rewrite and with few regressions, announce the EOL of our sdl1 backend at the same time and then proceed to rip it out as soon as it is convenient. I don't think we can get subviews to release quality in a month and a half, otherwise I'd propose it for that.

Thoughts?

**Comment by bradallred** [comment]
we might as well wait till subviews is merged. The SDL 2 driver on master is horrible performance wise and doesn't offer much in terms of benefit. I also don't see a point in removing it over just letting it bit rot and accept contributions from anybody willing to submit them a la SDL itself is doing.

That said it certainly shouldn't be default once subviews is merged (and should probably issue a deprecated warning when building). There is at least one platform (Mac) where the performance of SDL 1.2 is completely unplayable and I expect this situation to continue to degrade (for all platforms) in more ways than over time.

**Comment by MarcelHB** [comment]
No general objection.

We should only find some way to make SDL2 and subviews code work without the special OpenGL stuff of #441, even at the cost of some visual artifcats, in a release distribution since this will otherwise exclude some comparably young platforms from running (#810), given the age of the original games.

Other than that: Easier CMake, no dual-maintenance, so I welcome that.

**Comment by bradallred** [comment]
> We should only find some way to make SDL2 and subviews code work without the special OpenGL 

that's not going to happen without writing a shader for every graphics backend you want to support. its simply not possible to do in a way that doesn't both require tons of code and rewriting of things and also at the significant expense of performance. At that level of work you might as well introduce a new backend that extends the SDL 2 backend with SDL_gfx (assuming that actually is capable of what we need) or another cross platform abstraction layer with the capabilities required.

**Comment by MarcelHB** [comment]
No, I just think of opt-out'ing the shader code at runtime if the alternative is not being able to play at all, at the cost of living with things like #775.

**Comment by bradallred** [comment]
well now I dont know what you are talking about. I can use SDL2 without OpenGL just fine as long as you accept that stenciling is gimped

**Comment by bradallred** [comment]
also I meant SDL_gpu, not SDL_gfx

**Comment by MarcelHB** [comment]
But this depends on a macro definition atm. With building a release distribution, we'd likely set `USE_OPENGL` for the best user experience -- or remove it and make it implicit. This, for example, makes some semi-old Intel GPUs on Linux excluded unless you recompile the game on your own, without the setting.

So I just propose to make this configurable. "Disable shaders" if your device is unable to run OpenGL 2.0 but SDL2 can do software rendering or whatever it has as a fallback. When dropping SDL1.2, #810 could use this as an easy workaround.

**Comment by bradallred** [comment]
ok, I understand you now. You should open an enhancement request, we can probably throw something together without too much hassle. If we just put the OpenGL stuff in its own DSO then we can attempt to load that at runtime if it fails to load due to missing dependancies we fallback.

**Comment by bradallred** [comment]
In there short term, there is nothing stopping us from distributing 2 different drivers

**Comment by lynxlynxlynx** [comment]
Ok, so we're roughly in accord — good.

Considering we apparently don't have a good safety net, leaving the backend in sounds like a good idea. People run gemrb on crays and other crazy equipment that sometimes definitely does not have any GL capabilities.

Once we switch to sdl2 as the default, we could always build it twice and then prefer the GL capable version when starting up. If it doesn't work, fall back to the other one. Would require some build system ugliness, but should be possible.

**Comment by MarcelHB** [comment]
I volunteer to look at this when I find time the next days.

**Comment by lynxlynxlynx** [comment]
What's the plan, https://github.com/gemrb/gemrb/issues/821#issuecomment-644348293 ?

**Comment by MarcelHB** [comment]
I'd say we default into building with OpenGL. Not sure if it's even worth looking at disabling our extensions, or switching to other SDL backends (D3D) on config requests. Last time I've checked is that SDL sw rendering is broken visually, and disabling our shader voodoo/switchting to some other backend has a notable performance penalty on slow devices (RPi 400).

And if this platform is somewhat too new for OpenGL (some SoCs afaik), there is Mesa3D zink as layer.

**Comment by MarcelHB** [comment]
Anyway, I can also leave this decision out of this, since I guess at some time in the future, we will maybe look at SDL2 side-by-side SDL3 stuff, and will have to choose between options again.

**Comment by lynxlynxlynx** [comment]
I have no problems with building with GL by default. Currently only some builders disable it, since we otherwise can't run some integration tests.

For slower devices the incompatibility with later GLES releases is more of an issue, at least from the feedback I got from some porters.

SDL3 is already in prerelease stages, but there is a bridge library like with the v1-v2 transition. So we can be v3 compatible without major changes.

**Comment by MarcelHB** [comment]
Then let's just simply get rid of SDL1.2 code/cmake first? We'll have a look at the other issues in following steps. Should be both orthogonal to this, and maybe easier to look at with less code.

**Comment by lynxlynxlynx** [comment]
Does it not work at all any more or are there just the graphical concession as before? If not so, I'd remove it (merge the work) after the next release, just announcing it before. But it would also mean we have no good software renderer, I guess.

**Comment by MarcelHB** [comment]
SDL1.2? I haven't touched that in years, so I don't know. Well, then this task can wait until then.

**Comment by lynxlynxlynx** [comment]
Adding it to the announcement already paid dividends. People are still using the SDL1 backend:

> I'm following this project for quite some time now, I'm using it to play BGT and I even ported, some months ago, GemRB 0.8.8 to AROS (AROS Research Operating System) an AmigaOS 3.1 compatible OS that runs on x86/x64/arm and m68K.
> I would like to ask if you don't mind, if SDL1 decommission is really going to happen.
> I'm asking that because on AROS there is no SDL2 at the moment (I don't know if this applies to other "exotic" platforms too) and SDL1 is working very well (I even had to default to sdlaudio since openal will crash gemrb sometimes on the x86 version).
> I'm going to try in the next months the porting of this version using my latest python 3.3 port, I hoped SDL1 could stay in place until BG2ee became "usable" ( playing bg2ee on a AmigaOS-like OS is something you don't see often :) ) so I'm trying to put in a good word for the "dear old" SDL1

So let's not rush it, even more than before.

**Comment by MarcelHB** [comment]
I'm more surprised it sill works.

**Comment by MarcelHB** [comment]
Spoiler: it doesn't work. I'm looking at it for trying some change wrt compatibility, but looking at the compiler errors, it likely hasn't worked for 2 years or something.

And well, it has a bunch of interesting visual artifacts by now.

<img width="642" height="507" alt="Image" src="https://github.com/user-attachments/assets/723d7f0f-2e8c-4296-a948-c2b4aa9b81d7" />
